File: src/assembler/Instruction.scala
package assembler

import data.Word32

/**
  * @author Christian Goldapp
  * @version 1.0
  */
abstract class Instruction(opcode: OpCode, a1: Byte, a2: Byte, a3: Byte, lineform: String) {
  def getBytes: Array[Byte]

  def getWords: Array[Word32] = Word32.bytesToWords(getBytes)

  def getOpCode = opcode

  def getWidth = 1

  override def toString = lineform
}

class NoWordInstruction(opcode: OpCode, arg1: Byte, arg2: Byte, arg3: Byte, lineform: String) extends Instruction(opcode: OpCode, arg1: Byte, arg2: Byte, arg3: Byte, lineform: String) {
  override def getBytes: Array[Byte] = Array(opcode.code, arg1, arg2, arg3)
  override def getWidth = 1
}

class OneWordInstruction(opcode: OpCode, arg1: Byte, arg2: Byte, arg3: Byte, w1: Word32, lineform: String) extends Instruction(opcode: OpCode, arg1: Byte, arg2: Byte, arg3: Byte, lineform: String) {

  override def getWords: Array[Word32] = Word32.bytesToWords(getBytes)

  override def getBytes: Array[Byte] = {
    val w1bytes = w1.bytes()
    Array(opcode.code, arg1, arg2, arg3, w1bytes(0), w1bytes(1), w1bytes(2), w1bytes(3))
  }

  override def getWidth = 2

}

class TwoWordInstruction(opcode: OpCode, arg1: Byte, arg2: Byte, arg3: Byte, w1: Word32, w2: Word32, lineform: String) extends Instruction(opcode: OpCode, arg1: Byte, arg2: Byte, arg3: Byte, lineform: String) {
  override def getBytes: Array[Byte] = {
    val w1bytes = w1.bytes()
    val w2bytes = w2.bytes()

    Array(opcode.code, arg1, arg2, arg3, w1bytes(0), w1bytes(1), w1bytes(2), w1bytes(3), w2bytes(0), w2bytes(1), w2bytes(2), w2bytes(3))
  }

  override def getWidth = 3
}

class ThreeWordInstruction(opcode: OpCode, arg1: Byte, arg2: Byte, arg3: Byte, w1: Word32, w2: Word32, w3: Word32, lineform: String) extends Instruction(opcode: OpCode, arg1: Byte, arg2: Byte, arg3: Byte, lineform: String) {
  override def getBytes: Array[Byte] = {
    val w1bytes = w1.bytes()
    val w2bytes = w2.bytes()
    val w3bytes = w3.bytes()

    Array(opcode.code, arg1, arg2, arg3, w1bytes(0), w1bytes(1), w1bytes(2), w1bytes(3), w2bytes(0), w2bytes(1), w2bytes(2), w2bytes(3), w3bytes(0), w3bytes(1), w3bytes(2), w3bytes(3))
  }

  override def getWidth = 4
}

//Placeholder for jumps as addresses are calculated later
class JumpInstruction(opcode: OpCode, arg1: Byte, arg2: Byte, label: String, jumpto: Int, w1: Word32, w2: Word32, lineform: String) extends TwoWordInstruction(opcode: OpCode, arg1, arg2, 0x00, w1, w2, lineform: String) {
  override def getBytes: Array[Byte] = Array(opcode.code, arg1, arg2, 0xFF.toByte)

  //This is the width used for the equivalent jump-to-address
  override def getWidth = {
    opcode match {
      case OpCode.JMP => 2
      case OpCode.JIZ => if (arg1 == 0xFF) 3 else 2
      case OpCode.JNZ => if (arg1 == 0xFF) 3 else 2
      case OpCode.JLZ => if (arg1 == 0xFF) 3 else 2
      case OpCode.JSZ => if (arg1 == 0xFF) 3 else 2
    }
  }
  def getJump = jumpto

  def getArg1 = arg1

  def getArg2 = arg2

  def getWord1 = w1

  def getWord2 = w2
}

File: src/assembler/Instruction.kt
package assembler

import common.ALL_HIGH
import common.OpCode
import common.Word32
import common.bytesToWords

sealed class Instruction(val opCode: OpCode, val arg1: Byte, val arg2: Byte, val arg3: Byte, val lineform: String) {
    abstract fun toBytes(): ByteArray
    fun toWords(): Array<Word32> = toBytes().bytesToWords()
    abstract val width: Int
    override fun toString() = lineform
}

class NoWordInstruction(opCode: OpCode, arg1: Byte, arg2: Byte, arg3: Byte, lineform: String) : Instruction(opCode, arg1, arg2, arg3, lineform) {
    override fun toBytes(): ByteArray = byteArrayOf(opCode.code, arg1, arg2, arg3)
    override val width = 1
}

class OneWordInstruction(opCode: OpCode, arg1: Byte, arg2: Byte, arg3: Byte, val w1: Word32, lineform: String) : Instruction(opCode, arg1, arg2, arg3, lineform) {
    override fun toBytes() = byteArrayOf(opCode.code, arg1, arg2, arg3, *w1.bytes())
    override val width = 2
}

open class TwoWordInstruction(opcode: OpCode, arg1: Byte, arg2: Byte, arg3: Byte, val w1: Word32, val w2: Word32, lineform: String) : Instruction(opcode, arg1, arg2, arg3, lineform) {
    override fun toBytes() = byteArrayOf(opCode.code, arg1, arg2, arg3, *w1.bytes(), *w2.bytes())
    override val width = 3
}

class ThreeWordInstruction(opCode: OpCode, arg1: Byte, arg2: Byte, arg3: Byte, val w1: Word32, val w2: Word32, val w3: Word32, lineform: String) : Instruction(opCode, arg1, arg2, arg3, lineform) {
    override fun toBytes() = byteArrayOf(opCode.code, arg1, arg2, arg3, *w1.bytes(), *w2.bytes(), *w3.bytes())
    override val width = 4
}

class JumpInstruction(opcode: OpCode, arg1: Byte, arg2: Byte, val label: String, val jumpto: Int, w1: Word32, w2: Word32, lineform: String) : TwoWordInstruction(opcode, arg1, arg2, 0x00, w1, w2, lineform) {
    override fun toBytes(): ByteArray = byteArrayOf(opCode.code, arg1, arg2, ALL_HIGH)
    override val width = when (opcode) {
        OpCode.JMP -> 2
        OpCode.JIZ -> if (arg1 == ALL_HIGH) 3 else 2
        OpCode.JNZ -> if (arg1 == ALL_HIGH) 3 else 2
        OpCode.JLZ -> if (arg1 == ALL_HIGH) 3 else 2
        OpCode.JSZ -> if (arg1 == ALL_HIGH) 3 else 2
        else -> -1
    }
}